sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs

```diff
@@ -60,16 +61,38 @@ private void Analyze(CSharpExplodedGraph explodedGraph, SyntaxNodeAnalysisContex
 
             try
             {
+                explodedGraph.ExplorationEnded += ExplorationEndedHandler;
+
                 explodedGraph.Walk();
             }
             finally
             {
-                foreach (var diagnostic in analyzerContexts.SelectMany(analyzerContext => analyzerContext.GetDiagnostics()))
-                {
-                    context.ReportDiagnosticWhenActive(diagnostic);
-                }
+                explodedGraph.ExplorationEnded -= ExplorationEndedHandler;
+            }
+
+            // Some of the rules can return good results if the tree was only partially visited; others need to completely
+            // walk the tree in order to avoid false positives.
+            //
+            // Due to this we split the rules in two sets and report the diagnostics in steps:
+            // - when the tree is successfully visited and ExplorationEnded event is raised
+            // - when the tree visit ends (explodedGraph.Walk() returns)
+            ReportDiagnostics(analyzerContexts, context, true);
+
+            analyzerContexts.ForEach(analyzerContext => analyzerContext.Dispose());
+
+            void ExplorationEndedHandler(object sender, EventArgs args)
+            {
+                ReportDiagnostics(analyzerContexts, context, false);
+            }
+        }
 
-                analyzerContexts.ForEach(analyzerContext => analyzerContext.Dispose());
+        private static void ReportDiagnostics(IEnumerable<ISymbolicExecutionAnalysisContext> analyzerContexts, SyntaxNodeAnalysisContext context, bool supportsPartialWalking)
+        {
+            foreach (var diagnostic in analyzerContexts
```

Comment:
hm do you know why the coverage says 3/4 conditions covered for the `foreach`? - see https://sonarcloud.io/code?id=sonaranalyzer-dotnet&pullRequest=3352&selected=sonaranalyzer-dotnet%3Asonaranalyzer-dotnet%2Fsrc%2FSonarAnalyzer.CSharp%2FRules%2FSymbolicExecution%2FSymbolicExecutionRunner.cs&line=91

Reply:
Hard to say. In the open cover report there are indeed 4 branches and one is not covered (line 91 from column 23 to 47).

```
<BranchPoints>
	<BranchPoint vc="1077" uspid="15803" ordinal="0" offset="23" sl="91" path="0" offsetend="25" fileid="1145" />
	<BranchPoint vc="0" uspid="15804" ordinal="1" offset="23" sl="91" path="1" offsetend="47" fileid="1145" />
	<BranchPoint vc="1" uspid="15805" ordinal="2" offset="58" sl="91" path="0" offsetend="60" fileid="1145" />
	<BranchPoint vc="1076" uspid="15806" ordinal="3" offset="58" sl="91" path="1" offsetend="83" fileid="1145" />
</BranchPoints>
```

This is what report generator is showing:
![image](https://user-images.githubusercontent.com/56015273/82822215-b3b50980-9ea5-11ea-9fe7-1cc26d37e394.png)

AFAICT it is related to generated code and I think it's not worth to cover just for the sake of having 100%.

Reply:
no, it was just out of curiosity